codec/reflectcodec/type_codec.go

```diff
@@ -607,44 +611,46 @@ func (c *genericCodec) unmarshal(p *wrappers.Packer, value reflect.Value, maxSli
 			)
 		}
 
-		numElts := int(numElts32)
-		valueTypeReflection := value.Type()
-		mapKeyType := valueTypeReflection.Key()
-		mapValueType := valueTypeReflection.Elem()
-
-		value.Set(reflect.MakeMapWithSize(valueTypeReflection, numElts))
+		var (
+			numElts      = int(numElts32)
+			mapType      = value.Type()
+			mapKeyType   = mapType.Key()
+			mapValueType = mapType.Elem()
+			prevKey      []byte
+		)
 
-		var lastKey []byte
+		// Set [value] to be a new map of the appropriate type.
+		value.Set(reflect.MakeMapWithSize(mapType, numElts))
 
 		for i := 0; i < numElts; i++ {
 			mapKey := reflect.New(mapKeyType).Elem()
-			mapValue := reflect.New(mapValueType).Elem()
 
-			keyStartPosition := p.Offset
+			keyStartOffset := p.Offset
 
 			if err := c.unmarshal(p, mapKey, c.maxSliceLen); err != nil {
 				return fmt.Errorf("couldn't unmarshal map key (%s): %w", mapKeyType, err)
 			}
 
-			// Get the key value from the slice of bytes and make sure, to be
-			// used in the next statement, if lastKey is available, and check
-			// the new key is actually bigger (according to bytes.Compare) than
+			// Get the key's byte representation and check that the new key
+			// is actually bigger (according to bytes.Compare) than
 			// the previous key.
 			//
-			// The reasoning is to discard any unsorted key, because the part of
-			// the serialization spec is to sort map keys before serializing,
-			// for consistency.
+			// We do this to enforce that key-value pairs are sorted by increasing key.
 			//
 			// See for context: https://github.com/ava-labs/avalanchego/pull/1790#discussion_r1283656558
-			keyBytes := p.Bytes[keyStartPosition:p.Offset]
-			if lastKey != nil && bytes.Compare(keyBytes, lastKey) < 0 {
-				return fmt.Errorf("keys are in the incorrect order (%s, %s)", lastKey, mapKey)
+			keyBytes := p.Bytes[keyStartOffset:p.Offset]
+			if i != 0 && bytes.Compare(keyBytes, prevKey) <= 0 {
```

Comment:
I think checking for `i != 0` (i.e. this isn't the first iteration) is a little more readable than checking for `lastKey != nil`

Reply:
Also, we want to enforce that `keyBytes` is _strictly larger_ than `prevKey` (i.e. we need <=, not <`